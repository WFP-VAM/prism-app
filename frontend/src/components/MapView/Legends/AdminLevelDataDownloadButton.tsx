import { mapValues } from 'lodash';
import moment from 'moment';
import React from 'react';
import { useSelector } from 'react-redux';
import { AdminLevelDataLayerProps } from '../../../config/types';
import { LayerData } from '../../../context/layers/layer-data';
import {
  dateRangeSelector,
  layerDataSelector,
} from '../../../context/mapStateSlice/selectors';
import { useSafeTranslation } from '../../../i18n';
import { castObjectsArrayToCsv } from '../../../utils/csv-utils';
import { DEFAULT_DATE_FORMAT_SNAKE_CASE } from '../../../utils/name-utils';
import MultiOptionsButton from '../../Common/MultiOptionsButton';
import { downloadToFile } from '../utils';

interface IProps {
  layer: AdminLevelDataLayerProps;
}

function AdminLevelDataDownloadButton({ layer }: IProps) {
  const { startDate: selectedDate } = useSelector(dateRangeSelector);
  const adminLevelLayerData = useSelector(
    layerDataSelector(layer.id, selectedDate),
  ) as LayerData<AdminLevelDataLayerProps>;

  const { t } = useSafeTranslation();

  const getFilename = (): string => {
    const safeTitle = layer.title ?? layer.id;
    if (selectedDate && layer.dates) {
      const dateString = moment(selectedDate).format(
        DEFAULT_DATE_FORMAT_SNAKE_CASE,
      );
      return `${safeTitle}_${dateString}`;
    }
    return safeTitle;
  };

  const handleDownloadGeoJson = (): void => {
    downloadToFile(
      {
        content: JSON.stringify(adminLevelLayerData.data.features),
        isUrl: false,
      },
      getFilename(),
      'application/json',
    );
  };

  const handleDownloadCsv = (): void => {
    const translatedColumnsNames = mapValues(
      adminLevelLayerData.data.layerData[0],
      (v, k) => (k === 'value' ? t(adminLevelLayerData.layer.id) : t(k)),
    );
    downloadToFile(
      {
        content: castObjectsArrayToCsv(
          adminLevelLayerData.data.layerData,
          translatedColumnsNames,
          ';',
        ),
        isUrl: false,
      },
      getFilename(),
      'text/csv',
    );
  };

  return (
    <MultiOptionsButton
      mainLabel={t('Download')}
      options={[
        {
          label: 'GEOJSON',
          onClick: handleDownloadGeoJson,
        },
        {
          label: 'CSV',
          onClick: handleDownloadCsv,
        },
      ]}
    />
  );
}

export default AdminLevelDataDownloadButton;
