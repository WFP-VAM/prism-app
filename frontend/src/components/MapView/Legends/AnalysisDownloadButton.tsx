import React, { useCallback, useMemo } from 'react';
import { useSelector } from 'react-redux';
import {
  analysisResultSelector,
  analysisResultSortByKeySelector,
  analysisResultSortOrderSelector,
} from '../../../context/analysisResultStateSlice';
import { useSafeTranslation } from '../../../i18n';
import {
  BaselineLayerResult,
  downloadCSVFromTableData,
  ExposedPopulationResult,
  generateAnalysisFilename,
  PolygonAnalysisResult,
  useAnalysisTableColumns,
} from '../../../utils/analysis-utils';
import MultiOptionsButton from '../../Common/MultiOptionsButton';
import { downloadToFile } from '../utils';

function AnalysisDownloadButton() {
  const analysisResult = useSelector(analysisResultSelector);
  const analysisResultSortByKey = useSelector(analysisResultSortByKeySelector);
  const analysisResultSortOrder = useSelector(analysisResultSortOrderSelector);
  const { translatedColumns } = useAnalysisTableColumns(analysisResult);

  const { t } = useSafeTranslation();

  const featureCollection = useMemo(() => {
    return analysisResult?.featureCollection;
  }, [analysisResult]);

  const doesLayerAcceptCSVDownload = useMemo(() => {
    return (
      analysisResult &&
      (analysisResult instanceof BaselineLayerResult ||
        analysisResult instanceof PolygonAnalysisResult)
    );
  }, [analysisResult]);

  const analysisDate = useMemo(() => {
    if (analysisResult instanceof BaselineLayerResult) {
      return analysisResult.analysisDate;
    }
    if (analysisResult instanceof PolygonAnalysisResult) {
      return analysisResult.endDate;
    }
    return null;
  }, [analysisResult]);

  const fileName = useMemo(() => {
    if (
      // Explicit condition for type narrowing
      !analysisResult ||
      analysisResult instanceof ExposedPopulationResult
    ) {
      return analysisResult?.getTitle(t);
    }
    return generateAnalysisFilename(analysisResult, analysisDate ?? null);
  }, [analysisDate, analysisResult, t]);

  const handleAnalysisDownloadGeoJson = useCallback((): void => {
    downloadToFile(
      {
        content: JSON.stringify(featureCollection),
        isUrl: false,
      },
      fileName ?? 'prism_extract',
      'application/json',
    );
  }, [featureCollection, fileName]);

  const handleAnalysisDownloadCsv = useCallback((): void => {
    if (
      // Explicit condition for type narrowing
      !analysisResult ||
      analysisResult instanceof ExposedPopulationResult
    ) {
      return;
    }
    downloadCSVFromTableData(
      analysisResult,
      translatedColumns,
      analysisDate ?? null,
      analysisResultSortByKey,
      analysisResultSortOrder,
    );
  }, [
    analysisDate,
    analysisResult,
    analysisResultSortByKey,
    analysisResultSortOrder,
    translatedColumns,
  ]);

  return (
    <MultiOptionsButton
      mainLabel={t('Download')}
      options={[
        {
          label: 'GEOJSON',
          onClick: handleAnalysisDownloadGeoJson,
        },
        ...(doesLayerAcceptCSVDownload
          ? [
              {
                label: 'CSV',
                onClick: handleAnalysisDownloadCsv,
              },
            ]
          : []),
      ]}
    />
  );
}

export default AnalysisDownloadButton;
