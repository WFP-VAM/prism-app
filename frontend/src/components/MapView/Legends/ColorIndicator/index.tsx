import React from 'react';
import { createStyles, makeStyles, Typography } from '@material-ui/core';
import { fade } from '@material-ui/core/styles/colorManipulator';

interface StylesProps {
  color: string;
  opacity: number;
  fillPattern: 'left' | 'right' | undefined;
}

const indicatorsCommonStyles = {
  height: 10,
  width: 10,
  marginRight: 4,
};

const useStyles = makeStyles(() =>
  createStyles({
    container: {
      display: 'flex',
      alignItems: 'center',
    },
    indicator: {
      ...indicatorsCommonStyles,
      backgroundColor: ({ color, opacity }: StylesProps) =>
        fade(color, opacity),
    },
    fillPatternIndicator: {
      ...indicatorsCommonStyles,
      background: ({ color, fillPattern }: StylesProps) =>
        `repeating-linear-gradient(to ${fillPattern} bottom, ${color}, ${color} 2px, white 2px, white 4px)`,
    },
  }),
);

function ColorIndicator({
  value,
  color,
  opacity,
  fillPattern,
}: ColorIndicatorProps) {
  const classes = useStyles({ color, opacity, fillPattern });

  return (
    <div className={classes.container}>
      <div
        className={
          fillPattern ? classes.fillPatternIndicator : classes.indicator
        }
      />
      <Typography color="textSecondary">{value}</Typography>
    </div>
  );
}

export interface ColorIndicatorProps {
  value: string;
  color: string;
  opacity: number;
  fillPattern?: 'left' | 'right';
}

export default ColorIndicator;
