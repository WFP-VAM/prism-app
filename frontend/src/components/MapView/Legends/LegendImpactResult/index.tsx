import React, { memo } from 'react';
import { useSafeTranslation } from '../../../../i18n';

interface LegendImpactProps {
  legendText: string;
  thresholdAbove?: number;
  thresholdBelow?: number;
}

const LegendImpactResult = memo(
  ({ legendText, thresholdBelow, thresholdAbove }: LegendImpactProps) => {
    const { t } = useSafeTranslation();
    return (
      <>
        {`${t('Impact Analysis on')}: ${t(legendText)}`}
        <br />
        {thresholdAbove ? `${t('Above Threshold')}: ${thresholdAbove}` : ''}
        <br />
        {thresholdBelow ? `${t('Below Threshold')}: ${thresholdBelow}` : ''}
      </>
    );
  },
);

export default LegendImpactResult;
