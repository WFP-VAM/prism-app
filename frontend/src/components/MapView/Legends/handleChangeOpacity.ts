import { Map as MaplibreMap } from 'maplibre-gl';
import { LayerType } from 'config/types';
import React from 'react';
import { getLayerMapId } from 'utils/map-utils';

export const handleChangeOpacity = (
  event: React.ChangeEvent<{}>,
  newValue: number,
  map: MaplibreMap | undefined,
  id: LayerType['id'] | undefined,
  type: LayerType['type'] | undefined,
  callback: (newVal: number) => void,
) => {
  // TODO: temporary solution for opacity adjustment, we hope to edit react-mapbox in the future to support changing props
  // because the whole map will be re-rendered if using state directly
  if (map && id) {
    const [layerId, opacityType] = ((
      layerType?: LayerType['type'],
    ): [string, string] => {
      switch (layerType) {
        case 'wms':
          return [getLayerMapId(id), 'raster-opacity'];
        case 'static_raster':
          return [getLayerMapId(id), 'raster-opacity'];
        case 'admin_level_data':
        case 'composite':
        case 'impact':
          return [getLayerMapId(id), 'fill-opacity'];
        case 'point_data':
          // This is a hacky way to support opacity change for Kobo data.
          // TODO - Handle Kobo data as admin_level_data instead of point_data. See issue #760.
          if (id?.includes('_report')) {
            return [getLayerMapId(id), 'fill-opacity'];
          }
          return [getLayerMapId(id), 'circle-opacity'];
        // analysis layer type is undefined TODO we should try make analysis a layer to remove edge cases like this
        case undefined:
          return ['layer-analysis', 'fill-opacity'];
        default:
          throw new Error('Unknown map layer type');
      }
    })(type);

    map.setPaintProperty(layerId, opacityType, newValue);
    // force a update of the map style to ensure the change is reflected
    // see https://github.com/maplibre/maplibre-gl-js/issues/3373
    // TODO - check if the above issue got resolved from time to time.
    // eslint-disable-next-line no-underscore-dangle
    map.style._updateLayer(layerId as any);
    callback(newValue);
  }
};
