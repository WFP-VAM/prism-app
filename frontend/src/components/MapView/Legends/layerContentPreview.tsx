import React, { useRef, useState, useLayoutEffect } from 'react';
import {
  IconButton,
  createStyles,
  Grid,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import InfoIcon from '@material-ui/icons/Info';
import { LayerType } from '../../../config/types';
import { LayerDefinitions } from '../../../config/utils';
import ContentDialog, { loadLayerContent } from '../../NavBar/ContentDialog';

const LayerContentPreview = ({ layerId, classes }: PreviewProps) => {
  const [content, setContent] = useState<string | undefined>(undefined);
  const contentRef = useRef<HTMLHeadingElement>(null);
  const layer = LayerDefinitions[layerId || 'admin_boundaries'];

  useLayoutEffect(() => {
    if (contentRef.current !== null) {
      contentRef.current.scrollIntoView({
        behavior: 'smooth',
        block: 'start',
      });
    }
  });

  if (!layer.contentPath) {
    return null;
  }

  return (
    <Grid item>
      <IconButton size="small" className={classes.icon}>
        <InfoIcon
          fontSize="inherit"
          onClick={() => loadLayerContent(layer.contentPath!, setContent)}
        />
      </IconButton>
      <ContentDialog content={content} setContent={setContent} />
    </Grid>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    icon: {
      top: '-4px',
      fontSize: '14px',
    },
    label: {
      marginLeft: '10px',
    },
    title: {
      color: theme.palette.text.secondary,
      fontWeight: 'bold',
      minWidth: '600px',
    },
  });

export interface PreviewProps extends WithStyles<typeof styles> {
  layerId?: LayerType['id'];
}

export default withStyles(styles)(LayerContentPreview);
