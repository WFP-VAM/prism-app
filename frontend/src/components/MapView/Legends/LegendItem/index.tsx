import React, {
  memo,
  PropsWithChildren,
  useCallback,
  useEffect,
  useState,
  useMemo,
} from 'react';
import {
  Box,
  createStyles,
  Divider,
  Grid,
  IconButton,
  ListItem,
  Paper,
  Popover,
  Slider,
  Tooltip,
  Typography,
  withStyles,
  WithStyles,
} from '@material-ui/core';
import { Close, Opacity } from '@material-ui/icons';
import { useDispatch, useSelector } from 'react-redux';
import { LayerType, LegendDefinitionItem } from 'config/types';
import { mapSelector, layersSelector } from 'context/mapStateSlice/selectors';
import { clearDataset } from 'context/datasetStateSlice';
import { useSafeTranslation } from 'i18n';
import { clearAnalysisResult } from 'context/analysisResultStateSlice';
import LayerContentPreview from 'components/MapView/Legends/layerContentPreview';
import ColorIndicator from 'components/MapView/Legends/ColorIndicator';
import { getLegendItemLabel } from 'components/MapView/utils';
import { Extent } from 'components/MapView/Layers/raster-utils';
import { getUrlKey, useUrlHistory } from 'utils/url-utils';
import LayerDownloadOptions from 'components/MapView/LeftPanel/layersPanel/MenuItem/MenuSwitch/SwitchItem/LayerDownloadOptions';
import AnalysisDownloadButton from 'components/MapView/Legends//AnalysisDownloadButton';
import { toggleRemoveLayer } from 'components/MapView/LeftPanel/layersPanel/MenuItem/MenuSwitch/SwitchItem/utils';
import { opacitySelector, setOpacity } from 'context/opacityStateSlice';
import LoadingBar from '../LoadingBar';

// Children here is legendText
const LegendItem = memo(
  ({
    classes,
    id,
    title,
    legend,
    type,
    opacity: initialOpacity,
    children,
    legendUrl,
    fillPattern,
    extent,
  }: LegendItemProps) => {
    const dispatch = useDispatch();
    const { removeLayerFromUrl } = useUrlHistory();
    const map = useSelector(mapSelector);
    const [opacityEl, setOpacityEl] = useState<HTMLButtonElement | null>(null);
    const opacity = useSelector(opacitySelector(id as string));
    const isAnalysis = type === 'analysis';

    useEffect(() => {
      if (opacity !== undefined) {
        return;
      }
      dispatch(
        setOpacity({
          map,
          value: initialOpacity || 0,
          layerId: id,
          layerType: type,
        }),
      );
    }, [dispatch, id, initialOpacity, map, opacity, type]);

    const { t } = useSafeTranslation();

    const openOpacity = (event: React.MouseEvent<HTMLButtonElement>) => {
      setOpacityEl(event.currentTarget);
    };

    const closeOpacity = () => {
      setOpacityEl(null);
    };

    const open = Boolean(opacityEl);
    const opacityId = open ? 'opacity-popover' : undefined;

    const selectedLayers = useSelector(layersSelector);
    const layer = useMemo(() => {
      return selectedLayers.find(l => l.id === id);
    }, [id, selectedLayers]);

    const renderedOpacitySlider = useMemo(() => {
      return (
        <Box px={2} display="flex" className={classes.opacityBox}>
          <Typography classes={{ root: classes.opacityText }}>
            {`${Math.round((opacity || 0) * 100)}%`}
          </Typography>
          <Slider
            value={opacity}
            step={0.01}
            min={0}
            max={1}
            aria-labelledby="opacity-slider"
            classes={{
              root: classes.opacitySliderRoot,
              thumb: classes.opacitySliderThumb,
            }}
            onChange={(e, newValue) =>
              dispatch(
                setOpacity({
                  map,
                  value: newValue as number,
                  layerId: id,
                  layerType: type,
                }),
              )
            }
          />
        </Box>
      );
    }, [
      classes.opacityBox,
      classes.opacitySliderRoot,
      classes.opacitySliderThumb,
      classes.opacityText,
      dispatch,
      id,
      map,
      opacity,
      type,
    ]);

    const layerDownloadOptions = useMemo(() => {
      return layer ? (
        <LayerDownloadOptions
          layerId={layer.id}
          extent={extent}
          selected
          size="small"
        />
      ) : null;
    }, [layer, extent]);

    const remove = useCallback(() => {
      if (isAnalysis) {
        dispatch(clearAnalysisResult());
      }
      if (layer) {
        // clear previous table dataset loaded first
        // to close the dataseries and thus close chart
        dispatch(clearDataset());
        const urlLayerKey = getUrlKey(layer);
        toggleRemoveLayer(
          layer,
          map,
          urlLayerKey,
          dispatch,
          removeLayerFromUrl,
        );
      }
    }, [isAnalysis, layer, dispatch, map, removeLayerFromUrl]);

    const getColorIndicatorKey = useCallback((item: LegendDefinitionItem) => {
      return (
        item.value ||
        (typeof item.label === 'string' ? item?.label : item?.label?.text)
      );
    }, []);

    const renderedLegendDefinitionItems = useMemo(() => {
      return legend?.map((item: LegendDefinitionItem) => (
        <ColorIndicator
          key={getColorIndicatorKey(item)}
          value={getLegendItemLabel(t, item)}
          color={item.color as string}
          opacity={opacity as number}
          fillPattern={fillPattern}
        />
      ));
    }, [fillPattern, getColorIndicatorKey, legend, opacity, t]);

    const renderedLegendUrl = useMemo(() => {
      if (legendUrl) {
        return <img src={legendUrl} alt={title} />;
      }
      return <>{renderedLegendDefinitionItems}</>;
    }, [legendUrl, renderedLegendDefinitionItems, title]);

    const renderedLegend = useMemo(() => {
      if (!legend) {
        return null;
      }
      return <Grid item>{renderedLegendUrl}</Grid>;
    }, [legend, renderedLegendUrl]);

    const renderedChildren = useMemo(() => {
      if (!children) {
        return null;
      }
      return (
        <Grid item>
          <Typography variant="h5">{children}</Typography>
        </Grid>
      );
    }, [children]);

    return (
      <ListItem disableGutters dense>
        <Paper className={classes.paper}>
          <Grid item style={{ display: 'flex' }}>
            <Typography style={{ flexGrow: 1 }} variant="h4">
              {title}
            </Typography>
            <LayerContentPreview layerId={id} />
          </Grid>
          <Divider />
          {renderedLegend}
          <LoadingBar layerId={id} />
          {renderedChildren}
          <Divider style={{ margin: '8px 0px' }} />
          <Box display="flex" justifyContent="space-between">
            <Tooltip title="Opacity">
              <IconButton size="small" onClick={openOpacity}>
                <Opacity fontSize="small" />
              </IconButton>
            </Tooltip>
            <>
              <Popover
                id={opacityId}
                open={open}
                anchorEl={opacityEl}
                onClose={closeOpacity}
                anchorOrigin={{
                  vertical: 'bottom',
                  horizontal: 'left',
                }}
              >
                {renderedOpacitySlider}
              </Popover>
              {isAnalysis ? <AnalysisDownloadButton /> : layerDownloadOptions}
              <Tooltip title="Remove layer">
                <IconButton size="small" onClick={remove}>
                  <Close fontSize="small" />
                </IconButton>
              </Tooltip>
            </>
          </Box>
        </Paper>
      </ListItem>
    );
  },
);

const styles = () =>
  createStyles({
    paper: {
      padding: 8,
      width: 180,
    },
    slider: {
      padding: '0 5px',
    },
    opacityBox: {
      backgroundColor: '#fff',
      width: 172,
      overflow: 'hidden',
    },
    opacitySliderRoot: {
      color: '#4CA1AD',
      flexGrow: 1,
      padding: '18px 0',
    },
    opacitySliderThumb: {
      backgroundColor: '#4CA1AD',
    },
    opacityText: {
      color: '#4CA1AD',
      marginRight: 5,
      width: 28,
      lineHeight: '36px',
    },
  });

interface LegendItemProps
  extends WithStyles<typeof styles>,
    PropsWithChildren<{}> {
  id: LayerType['id'];
  title: LayerType['title'];
  legend: LayerType['legend'];
  legendUrl?: string;
  type: LayerType['type'] | 'analysis';
  opacity: LayerType['opacity'];
  fillPattern?: 'left' | 'right';
  extent?: Extent;
}

export default withStyles(styles)(LegendItem);
