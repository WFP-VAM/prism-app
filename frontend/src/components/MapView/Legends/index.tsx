import {
  Button,
  createStyles,
  Divider,
  Grid,
  Hidden,
  List,
  Typography,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { Visibility, VisibilityOff } from '@material-ui/icons';
import React, { useState, memo, useMemo, useCallback } from 'react';

import { createGetLegendGraphicUrl } from 'prism-common';
import { useSelector } from 'react-redux';
import {
  analysisResultOpacitySelector,
  analysisResultSelector,
  isAnalysisLayerActiveSelector,
} from 'context/analysisResultStateSlice';
import { LayerType } from 'config/types';
import { BaselineLayerResult } from 'utils/analysis-utils';
import { useSafeTranslation } from 'i18n';

import AnalysisDownloadButton from './AnalysisDownloadButton';
import LegendItem from './LegendItem';
import LegendImpactResult from './LegendImpactResult';

const Legends = memo(({ classes, layers }: LegendsProps) => {
  // Selectors
  const isAnalysisLayerActive = useSelector(isAnalysisLayerActiveSelector);
  const analysisResult = useSelector(analysisResultSelector);
  const analysisLayerOpacity = useSelector(analysisResultOpacitySelector);

  // State
  const [open, setOpen] = useState(true);

  // memoized values from selectors
  const featureCollection = useMemo(() => {
    return analysisResult?.featureCollection;
  }, [analysisResult]);

  const hasData = useMemo(() => {
    return featureCollection?.features
      ? featureCollection.features.length > 0
      : false;
  }, [featureCollection]);

  const { t } = useSafeTranslation();

  // If legend array is empty, we fetch from remote server the legend as GetLegendGraphic request.
  const getLayerLegendUrl = useCallback((layer: LayerType) => {
    return layer.type === 'wms' && layer.legend.length === 0
      ? createGetLegendGraphicUrl({
          base: layer.baseUrl,
          layer: layer.serverLayerName,
        })
      : undefined;
  }, []);

  const layersLegendItems = useMemo(() => {
    return layers.map(layer => {
      if (!layer.legend || !layer.legendText) {
        // this layer doesn't have a legend (likely boundary), so lets ignore.
        return null;
      }
      return (
        <LegendItem
          key={layer.id}
          id={layer.id}
          title={layer.title ? t(layer.title) : undefined}
          legend={layer.legend}
          legendUrl={getLayerLegendUrl(layer)}
          type={layer.type}
          opacity={layer.opacity}
          fillPattern={layer.fillPattern}
        >
          {t(layer.legendText)}
        </LegendItem>
      );
    });
  }, [getLayerLegendUrl, layers, t]);

  const renderedLegendImpactResult = useMemo(() => {
    if (!(analysisResult instanceof BaselineLayerResult)) {
      return null;
    }
    const baselineLayer = analysisResult.getBaselineLayer();
    const hazardLayer = analysisResult.getHazardLayer();
    return (
      <LegendImpactResult
        legendText={
          baselineLayer.legendText
            ? baselineLayer.legendText
            : hazardLayer.legendText
        }
        thresholdBelow={analysisResult.threshold.below}
        thresholdAbove={analysisResult.threshold.above}
      />
    );
  }, [analysisResult]);

  // add analysis legend item if layer is active and analysis result exists
  const analysisLegendItem = useMemo(() => {
    if (!isAnalysisLayerActive || !hasData) {
      return [];
    }
    return [
      <LegendItem
        key={analysisResult?.key ?? Date.now()}
        legend={analysisResult?.legend}
        title={analysisResult?.getTitle(t)}
        opacity={analysisLayerOpacity} // TODO: initial opacity value
        // Control opacity only for analysis
        // for the other layers it is controlled from the left panel
        displayOpacitySlider={isAnalysisLayerActive && hasData}
      >
        {renderedLegendImpactResult}
        <Divider />
        <Grid item>
          <AnalysisDownloadButton />
        </Grid>
      </LegendItem>,
    ];
  }, [
    analysisLayerOpacity,
    analysisResult,
    hasData,
    isAnalysisLayerActive,
    renderedLegendImpactResult,
    t,
  ]);

  const legendItems = useMemo(() => {
    return [...layersLegendItems, ...analysisLegendItem];
  }, [analysisLegendItem, layersLegendItems]);

  const renderedVisibilityButton = useMemo(() => {
    if (open) {
      return <VisibilityOff fontSize="small" />;
    }
    return <Visibility fontSize="small" />;
  }, [open]);

  const renderedLegendItemsList = useMemo(() => {
    if (!open) {
      return null;
    }
    return <List className={classes.list}>{legendItems}</List>;
  }, [classes.list, legendItems, open]);

  const toggleLegendVisibility = useCallback(() => {
    setOpen(!open);
  }, [open]);

  return (
    <Grid item className={classes.container}>
      <Button
        className={classes.triggerButton}
        variant="contained"
        color="primary"
        onClick={toggleLegendVisibility}
      >
        {renderedVisibilityButton}
        <Hidden smDown>
          <Typography className={classes.label} variant="body2">
            {t('Legend')}
          </Typography>
        </Hidden>
      </Button>
      {renderedLegendItemsList}
    </Grid>
  );
});

const styles = () =>
  createStyles({
    container: {
      textAlign: 'right',
    },
    triggerButton: {
      height: '3em',
    },
    label: {
      marginLeft: '10px',
    },
    list: {
      overflowX: 'hidden',
      overflowY: 'auto',
      maxHeight: '70vh',
      position: 'absolute',
      right: '16px',
    },
  });

export interface LegendsProps extends WithStyles<typeof styles> {
  layers: LayerType[];
}

export default withStyles(styles)(Legends);
