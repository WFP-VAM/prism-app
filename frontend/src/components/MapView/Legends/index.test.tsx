import React from 'react';
import { render } from '@testing-library/react';

import { Provider } from 'react-redux';
import Legends from '.';
import { store } from '../../../context/store';

jest.mock('./ColorIndicator', () => 'mock-ColorIndicator');

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <Legends layers={[]} />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
