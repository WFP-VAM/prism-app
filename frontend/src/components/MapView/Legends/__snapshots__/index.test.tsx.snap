// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiGrid-root Legends-container-1 MuiGrid-item"
  >
    <mock-button
      color="primary"
      variant="contained"
    >
      <mock-visibilityoff
        fontsize="small"
      />
      <mock-hidden
        smdown="true"
      >
        <mock-typography
          classname="Legends-label-2"
          variant="body2"
        >
          Legend
        </mock-typography>
      </mock-hidden>
    </mock-button>
    <mock-list
      classname="Legends-list-3"
    />
  </div>
</div>
`;
